File: web/src/lib/__tests__/price.test.ts
import {
  MAX_AMOUNT,
  MIN_AMOUNT,
  amountFromInput,
  formatPrice,
  isCanonicalPrice,
  isValidPrice,
  parsePrice,
  sanitizeAmountInput,
  type PriceValue,
} from '@/lib/price'

/**
 * This table is the same one as `parseCases` in api/pkg/price/price_test.go,
 * and the two must agree. The three implementations of the grammar — this one,
 * pkg/price, and the mentors_price_chk regexp — cannot import each other, so a
 * divergence only ever shows up as a save that 400s with no field attached, or
 * as a row the catalog cannot classify.
 *
 * Anything added here goes in the Go table too.
 */
const parseCases: { raw: string; want: PriceValue | null; why?: string }[] = [
  { raw: 'Free', want: { kind: 'free', amount: null } },
  { raw: 'Negotiable', want: { kind: 'negotiable', amount: null } },

  { raw: '$1', want: { kind: 'fixed', amount: 1 }, why: 'MIN_AMOUNT' },
  { raw: '$1000', want: { kind: 'fixed', amount: 1000 }, why: 'MAX_AMOUNT, inclusive' },
  // Every distinct amount live in production on 2026-08-13.
  { raw: '$10', want: { kind: 'fixed', amount: 10 } },
  { raw: '$20', want: { kind: 'fixed', amount: 20 } },
  { raw: '$30', want: { kind: 'fixed', amount: 30 } },
  { raw: '$40', want: { kind: 'fixed', amount: 40 } },
  { raw: '$50', want: { kind: 'fixed', amount: 50 } },
  { raw: '$60', want: { kind: 'fixed', amount: 60 } },
  { raw: '$70', want: { kind: 'fixed', amount: 70 } },
  { raw: '$80', want: { kind: 'fixed', amount: 80 } },
  { raw: '$90', want: { kind: 'fixed', amount: 90 } },
  { raw: '$100', want: { kind: 'fixed', amount: 100 } },
  { raw: '$120', want: { kind: 'fixed', amount: 120 } },
  { raw: '$150', want: { kind: 'fixed', amount: 150 } },
  { raw: '$200', want: { kind: 'fixed', amount: 200 } },

  { raw: '', want: null, why: 'empty is what an unset column used to save as' },
  { raw: '$', want: null, why: 'the sigil alone' },
  { raw: '$0', want: null, why: 'a free session is Free, not zero dollars' },
  { raw: '$00', want: null, why: 'leading zero' },
  { raw: '$01', want: null, why: 'leading zero' },
  { raw: '$050', want: null, why: 'a second spelling of $50' },
  { raw: '$1001', want: null, why: 'one over the cap' },
  { raw: '$99999', want: null, why: 'far over the cap' },
  { raw: '$+50', want: null, why: 'Number() would accept the sign' },
  { raw: '$-50', want: null, why: 'Number() would accept the sign' },
  { raw: '50', want: null, why: "bare digits: what mapPrice's passthrough could emit" },
  { raw: '$50.00', want: null, why: 'no decimals' },
  { raw: '$1,000', want: null, why: 'no thousands separator' },
  { raw: '$50 / hour', want: null, why: 'the free-text shape D3 allowed' },
  { raw: '$100 / hour', want: null, why: 'the free-text shape D3 allowed' },
  { raw: '$ 50', want: null, why: 'space after the sigil' },
  { raw: ' Free', want: null, why: 'parsePrice does not trim' },
  { raw: 'Free ', want: null, why: 'parsePrice does not trim' },
  { raw: 'Free\n', want: null, why: 'a trailing newline is still not Free' },
  { raw: 'free', want: null, why: 'parsePrice does not case-fold' },
  { raw: 'FREE', want: null, why: 'parsePrice does not case-fold' },
  { raw: 'negotiable', want: null, why: 'parsePrice does not case-fold' },
  { raw: 'Negotiable price', want: null, why: 'the canonical word is the whole value' },
  { raw: '$٥٠', want: null, why: 'Arabic-Indic digits are not ASCII digits' },
  { raw: '€50', want: null, why: 'one currency only' },
]

describe('parsePrice', () => {
  it.each(parseCases)('parses $raw', ({ raw, want, why }) => {
    expect(parsePrice(raw)).toEqual(want)
    expect(isCanonicalPrice(raw)).toBe(want !== null)
    if (why !== undefined && want === null) {
      // Keeps the reason attached to the case in the failure output.
      expect(`${raw}: ${why}`).toContain(why)
    }
  })
})

describe('formatPrice', () => {
  it.each(parseCases.filter((c) => c.want !== null))(
    'round-trips $raw byte for byte',
    ({ raw, want }) => {
      expect(formatPrice(want as PriceValue)).toBe(raw)
    }
  )

  // Refuses to render a pair parsePrice could not have produced, so a form that
  // submits it writes a price the database constraint rejects rather than a
  // plausible-looking wrong one.
  it.each([
    [{ kind: 'fixed', amount: null }, 'no amount chosen yet'],
    [{ kind: 'fixed', amount: 0 }, 'below MIN_AMOUNT'],
    [{ kind: 'fixed', amount: 1001 }, 'above MAX_AMOUNT'],
    [{ kind: 'fixed', amount: -50 }, 'negative'],
    [{ kind: 'fixed', amount: 50.5 }, 'not whole'],
  ] as [PriceValue, string][])('returns null for %p (%s)', (value) => {
    expect(formatPrice(value)).toBeNull()
  })

  // Kind is the discriminator, not amount === 0: a stale amount left over from
  // switching away from "fixed" must not leak into the rendered value.
  it('ignores a stale amount on the amount-less kinds', () => {
    expect(formatPrice({ kind: 'free', amount: 120 })).toBe('Free')
    expect(formatPrice({ kind: 'negotiable', amount: 120 })).toBe('Negotiable')
  })
})

describe('isValidPrice', () => {
  it('accepts the amount-less kinds with no amount', () => {
    expect(isValidPrice({ kind: 'free', amount: null })).toBe(true)
    expect(isValidPrice({ kind: 'negotiable', amount: null })).toBe(true)
  })

  it('requires a whole in-range amount for fixed', () => {
    expect(isValidPrice({ kind: 'fixed', amount: null })).toBe(false)
    expect(isValidPrice({ kind: 'fixed', amount: 0 })).toBe(false)
    expect(isValidPrice({ kind: 'fixed', amount: 1001 })).toBe(false)
    expect(isValidPrice({ kind: 'fixed', amount: 50 })).toBe(true)
  })
})

describe('sanitizeAmountInput', () => {
  it.each([
    ['', '', 'an empty box stays empty'],
    ['50', '50', 'digits pass through'],
    ['abc', '', 'letters are not an amount'],
    ['5a0', '50', 'letters are dropped from between digits'],
    ['0', '', 'a lone zero clears rather than sitting there as an invalid amount'],
    ['007', '7', 'leading zeros are dropped'],
    ['$50', '50', 'the sigil is a prefix element, never part of the value'],
    ['50.5', '505', 'decimals are not a thing this grammar has'],
    ['-50', '50', 'a sign is not part of an amount'],
    ['1000', '1000', 'MAX_AMOUNT must stay typable'],
    ['12345', '1234', 'capped at the digits MAX_AMOUNT has'],
    ['  50  ', '50', 'whitespace is dropped'],
  ])('sanitizes %p to %p (%s)', (raw, want) => {
    expect(sanitizeAmountInput(raw)).toBe(want)
  })
})

describe('amountFromInput', () => {
  it('reports null for an empty box rather than 0', () => {
    expect(amountFromInput('')).toBeNull()
    expect(amountFromInput('abc')).toBeNull()
    expect(amountFromInput('0')).toBeNull()
  })

  it('reads the sanitized digits', () => {
    expect(amountFromInput('50')).toBe(50)
    expect(amountFromInput('$120')).toBe(120)
  })
})

describe('bounds', () => {
  it('forbids a zero price at the floor', () => {
    expect(MIN_AMOUNT).toBe(1)
    expect(MAX_AMOUNT).toBe(1000)
  })
})

File: api/pkg/price/price_test.go
package price_test

import (
	"testing"

	"github.com/openmentor-io/openmentor/api/pkg/price"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// parseCases is the definition of the grammar, and three implementations have
// to agree with it: this package, the mentors_price_chk regexp in
// api/migrations/000014_price_grammar.up.sql (proved against a real database by
// api/internal/repository/mentor_price_constraint_db_test.go) and the browser
// copy in web/src/lib/price.ts (pinned by web/src/lib/__tests__/price.test.ts,
// which carries this same table). A value one accepts and another rejects is a
// save that 400s with no field attached, or a row the catalog cannot classify.
var parseCases = []struct {
	raw    string
	want   price.Value
	valid  bool
	reason string
}{
	{raw: "Free", want: price.Value{Kind: price.Free}, valid: true},
	{raw: "Negotiable", want: price.Value{Kind: price.Negotiable}, valid: true},

	{raw: "$1", want: price.Value{Kind: price.Fixed, Amount: 1}, valid: true, reason: "MinAmount"},
	{raw: "$1000", want: price.Value{Kind: price.Fixed, Amount: 1000}, valid: true, reason: "MaxAmount, inclusive"},
	// Every distinct amount live in production on 2026-08-13.
	{raw: "$10", want: price.Value{Kind: price.Fixed, Amount: 10}, valid: true},
	{raw: "$20", want: price.Value{Kind: price.Fixed, Amount: 20}, valid: true},
	{raw: "$30", want: price.Value{Kind: price.Fixed, Amount: 30}, valid: true},
	{raw: "$40", want: price.Value{Kind: price.Fixed, Amount: 40}, valid: true},
	{raw: "$50", want: price.Value{Kind: price.Fixed, Amount: 50}, valid: true},
	{raw: "$60", want: price.Value{Kind: price.Fixed, Amount: 60}, valid: true},
	{raw: "$70", want: price.Value{Kind: price.Fixed, Amount: 70}, valid: true},
	{raw: "$80", want: price.Value{Kind: price.Fixed, Amount: 80}, valid: true},
	{raw: "$90", want: price.Value{Kind: price.Fixed, Amount: 90}, valid: true},
	{raw: "$100", want: price.Value{Kind: price.Fixed, Amount: 100}, valid: true},
	{raw: "$120", want: price.Value{Kind: price.Fixed, Amount: 120}, valid: true},
	{raw: "$150", want: price.Value{Kind: price.Fixed, Amount: 150}, valid: true},
	{raw: "$200", want: price.Value{Kind: price.Fixed, Amount: 200}, valid: true},

	{raw: "", reason: "empty is what an unset column used to save as"},
	{raw: "$", reason: "the sigil alone"},
	{raw: "$0", reason: "a free session is Free, not zero dollars"},
	{raw: "$00", reason: "leading zero"},
	{raw: "$01", reason: "leading zero"},
	{raw: "$050", reason: "a second spelling of $50"},
	{raw: "$1001", reason: "one over the cap"},
	{raw: "$99999", reason: "far over the cap"},
	{raw: "$+50", reason: "strconv.Atoi would accept the sign"},
	{raw: "$-50", reason: "strconv.Atoi would accept the sign"},
	{raw: "50", reason: "bare digits: what mapPrice's passthrough could emit"},
	{raw: "$50.00", reason: "no decimals"},
	{raw: "$1,000", reason: "no thousands separator"},
	{raw: "$50 / hour", reason: "the free-text shape D3 allowed"},
	{raw: "$100 / hour", reason: "the free-text shape D3 allowed"},
	{raw: "$ 50", reason: "space after the sigil"},
	{raw: " Free", reason: "Parse does not trim"},
	{raw: "Free ", reason: "Parse does not trim"},
	{raw: "Free\n", reason: "a trailing newline is still not Free"},
	{raw: "free", reason: "Parse does not case-fold"},
	{raw: "FREE", reason: "Parse does not case-fold"},
	{raw: "negotiable", reason: "Parse does not case-fold"},
	{raw: "Negotiable price", reason: "the canonical word is the whole value"},
	{raw: "$٥٠", reason: "Arabic-Indic digits are not ASCII digits"},
	{raw: "€50", reason: "one currency only"},
}

func TestParse(t *testing.T) {
	for _, tc := range parseCases {
		t.Run(tc.raw, func(t *testing.T) {
			got, err := price.Parse(tc.raw)

			if !tc.valid {
				assert.ErrorIs(t, err, price.ErrInvalid, "expected %q to be rejected (%s)", tc.raw, tc.reason)
				assert.False(t, price.IsCanonical(tc.raw))
				return
			}

			require.NoError(t, err, "expected %q to parse", tc.raw)
			assert.Equal(t, tc.want, got)
			assert.True(t, price.IsCanonical(tc.raw))
		})
	}
}

// A canonical string survives a Parse/String round trip byte for byte. This is
// what lets the API store what it validated instead of re-rendering it, and
// what will let the Option 3 backfill read the column with Parse and write it
// back with String without touching a single displayed value.
func TestParseStringRoundTrip(t *testing.T) {
	for _, tc := range parseCases {
		if !tc.valid {
			continue
		}
		t.Run(tc.raw, func(t *testing.T) {
			parsed, err := price.Parse(tc.raw)
			require.NoError(t, err)
			assert.Equal(t, tc.raw, parsed.String())
		})
	}
}

// String refuses to render a Value that Parse would not have produced, so a
// caller that built one by hand writes an empty price the database constraint
// rejects rather than a plausible-looking wrong one.
func TestStringRejectsValuesParseCannotProduce(t *testing.T) {
	assert.Equal(t, "", price.Value{}.String(), "the zero Value is not a price")
	assert.Equal(t, "", price.Value{Kind: "gratis"}.String(), "unknown kind")
	assert.Equal(t, "", price.Value{Kind: price.Fixed, Amount: 0}.String(), "below MinAmount")
	assert.Equal(t, "", price.Value{Kind: price.Fixed, Amount: 1001}.String(), "above MaxAmount")
	assert.Equal(t, "", price.Value{Kind: price.Fixed, Amount: -50}.String(), "negative")

	// Amount is ignored for the amount-less kinds rather than leaking into the
	// rendered value: Kind is the discriminator, not Amount == 0.
	assert.Equal(t, "Free", price.Value{Kind: price.Free, Amount: 50}.String())
	assert.Equal(t, "Negotiable", price.Value{Kind: price.Negotiable, Amount: 50}.String())
}

func TestBounds(t *testing.T) {
	assert.Equal(t, 1, price.MinAmount, "$0 must not be storable; a free session is Free")
	assert.Equal(t, 1000, price.MaxAmount)
}

// KindLabel is what the registration / profile-save / admin-update analytics
// events send as their bounded price dimension. Four values, closed set —
// derived from the same table so a grammar change moves it automatically.
func TestKindLabel(t *testing.T) {
	assert.Equal(t, "free", price.KindLabel("Free"))
	assert.Equal(t, "negotiable", price.KindLabel("Negotiable"))
	assert.Equal(t, "fixed", price.KindLabel("$137"))
	assert.Equal(t, "invalid", price.KindLabel("$50 / hour"))
	assert.Equal(t, "invalid", price.KindLabel(""))

	for _, tc := range parseCases {
		label := price.KindLabel(tc.raw)
		if tc.valid {
			assert.Equal(t, string(tc.want.Kind), label, "raw %q", tc.raw)
		} else {
			assert.Equal(t, "invalid", label, "raw %q", tc.raw)
		}
	}
}
